Preventing violations (#84)
AMWA-TV/nmos-sink-metadata-processing: docs/Behaviour - Server Side.md

Non-empty `constraint_sets` array MUST be treated as unordered unless at least one Constraint Set has `urn:x-nmos:cap:meta:preference` attribute.

> Active Constraints are satisfied when any of its Constraint Sets are satisfied in terms of [BCP-004-01][BCP-004-01].

Reviewer: I would keep this to make sure a Sender cannot advertise through  IS-04 a Flow or through IS-05 a transport file that violate the active constraints. If we simply state that a Sender cannot transmit to the network it keeps the possibility for a sender to advertise an illegal Flow and SDP transport file ...

"Once a Sender accepts proposed Active Constraints, this Sender, the Flow and the Connection API `/transportfile` resource (if used) associated with this Sender and the Source associated with this Flow MUST satisfy the Active Constraints when the Sender is active."

Author: I don't quite understand the point. A Sender has some Flow and transport file. Each time they change or Active Constraints change, the Node validates the Flow and the SDP against the Active Constraints. If validation fails, the Node prohibits transmitting with an illegal Flow and SDP. I don't understand, how we can avoid advertising an illegal Flow and SDP at all.

The removed paragraph doesn't save us from it as well because of "when the Sender is active" so I think "Preventing restrictions violation" in its current state covers this removed paragraph.

Reviewer: I would not want a Sender to be allowed to publish an associated Flow/Source that violate the active constraints. Just not transmitting to the network it not enough. This paragraph provide more control over such behavior. This is why I think we must keep it. What would be the reason to remove it?

Author: I removed it to have this language under "Preventing restrictions violation" and it seems to me to be covered there. 

> Once a Sender accepts proposed Active Constraints, <...> Flow <...> MUST satisfy the Active Constraints **when the Sender is active**.

For me it means that either the published Flow is illegal and Sender can't transmit it or it's legal and Sender can transmit it. This logic is described in "Preventing restrictions violation" now. Could you please explain how this paragraph prevents from publishing an illegal Flow?

Reviewer: Without this text "Once a Sender accepts proposed Active Constraints, this Sender, the Flow and the Connection API `/transportfile` resource (if used) associated with this Sender and the Source associated with this Flow MUST satisfy the Active Constraints when the Sender is active." the only restriction imposed by IS-11 is to not transmit a stream violating the active constraints over the network. IS-11 says nothing else. This is why I think it is important to include this text to add the additional restrictions that an active sender cannot publish an illegal flow ... only an inactive sender would be allowed.

Author: `Sender MUST NOT transmit` -> `Sender MUST become inactive` solved this issue, so I'm marking this as resolved.

> Active Constraints are satisfied when any of its Constraint Sets are satisfied in terms of [BCP-004-01][BCP-004-01].

Author: ```suggestion
Active Constraints are satisfied when any of its Constraint Sets are satisfied in terms of [BCP-004-01][BCP-004-01].
Active Constraints are violated when at least one Parameter Constraint of each Constraint Set is not satisfied in terms of [BCP-004-01][BCP-004-01].
```

## State of Sender

A Sender managed with IS-11 has the following states:
- `Unconstrained` when Active Constraints of this Sender is an empty `constraint_sets` array.

> - `Constrained` when the Active Constraints are satisfied.

Reviewer: "when at least one of the Active Constraints is satisfied."

Author: Active Constraints is an object with a single property "constraint_sets". I added "Active Constraints are satisfied when any of its Constraint Sets are satisfied in terms of [BCP-004-01][BCP-004-01]" to define what Active Constraints satisfaction means. If any properties are added to Active Constraints, it will be possible to change this single line instead of editing all the places referring to the satisfaction.

Reviewer: ok

> - `Active Constraints Violation` when at least one Parameter Constraint of each Constraint Set is not satisfied.

Reviewer: "when none of the Active Constraints is satisfied."

Author: Taking into account the comment above, should I add a definition of Active Constraints dissatisfaction and refer to it here?

Reviewer: Yes, would be more uniform ...

Author: ```suggestion
- `Active Constraints Violation` when Active Constraints are violated.
```

- `No Signal` when there is no signal from Inputs associated with this Sender.

> - `Awaiting Signal` when Inputs associated with this Sender were reconfigured recently and the signal is not stable yet. This is a transitional state until one of the previous ones can be established. It is RECOMMENDED to not update such associated resources as Flow and Source while this Sender is in this state.

Reviewer: "when the signal from Inputs associated to this Sender is transitioning and the signal is not stable yet. This is a transitional state until one of the previous ones can be established. It is RECOMMENDED to not update resources such as Flow and Source associated with a Sender while in this state.

Author: Great wording.

```suggestion
- `Awaiting Signal` when the signal from Inputs associated to this Sender is transitioning and the signal is not stable yet. This is a transitional state until one of the previous ones can be established. It is RECOMMENDED to not update resources such as Flow and Source associated with a Sender while in this state.
```

Reviewer: ok

## State of Receiver

A Receiver managed with IS-11 has the following states:
- `No Transport File` when there's no active transport file.
- `OK` when this Receiver's active transport file does not violate Receiver Capabilities.
- `Receiver Capabilities Violation` when this Receiver's active transport file violates Receiver Capabilities.

## Preventing restrictions violation

> At any time if State of an active Sender becomes `Active Constraints Violation`, the Sender MUST NOT transmit the Flow over the network. An inactive Sender in this state MUST NOT allow activations.

Author: ```suggestion
At any time if State of an active Sender becomes `Active Constraints Violation`, the Sender MUST become inactive. An inactive Sender in this state MUST NOT allow activations.
```

> At any time if State of an active Receiver becomes `Receiver Capabilities Violation`, the Receiver SHOULD stop receiving the Flow. An inactive Receiver in this state SHOULD NOT allow activations.

Author: ```suggestion
At any time if State of an active Receiver becomes `Receiver Capabilities Violation`, the Receiver SHOULD become inactive. An inactive Receiver in this state SHOULD NOT allow activations.
```

## State of Sender

A Sender managed with IS-11 has the following states:
- `Unconstrained` when Active Constraints of this Sender is an empty `constraint_sets` array.
- `Constrained` when the Active Constraints are satisfied.
- `Active Constraints Violation` when Active Constraints are violated.
- `No Signal` when there is no signal from Inputs associated with this Sender.
- `Awaiting Signal` when the signal from Inputs associated to this Sender is transitioning and the signal is not stable yet. This is a transitional state until one of the previous ones can be established. It is RECOMMENDED to not update resources such as Flow and Source associated with a Sender while in this state.

## State of Receiver

A Receiver managed with IS-11 has the following states:
- `No Transport File` when there's no active transport file.
- `OK` when this Receiver's active transport file does not violate Receiver Capabilities.
- `Receiver Capabilities Violation` when this Receiver's active transport file violates Receiver Capabilities.

## Preventing restrictions violation

> At any time if State of an active Sender becomes `Active Constraints Violation`, the Sender MUST become inactive. An inactive Sender in this state MUST NOT allow activations.

Reviewer: OK
